src/nodelet/line_segment_detector_nodelet.cpp

```diff
@@ -0,0 +1,254 @@
+#include <opencv2/imgproc/imgproc.hpp>
+
+#include <dynamic_reconfigure/server.h>
+#include "opencv_apps/Line.h"
+#include "opencv_apps/LineArrayStamped.h"
+#include "opencv_apps/LineSegmentDetectorConfig.h"
+
+
+namespace line_segment_detector {
+class LineSegmentDetectorNodelet : public opencv_apps::Nodelet
+{
+  image_transport::Publisher img_pub_;
+  image_transport::Subscriber img_sub_;
+  image_transport::CameraSubscriber cam_sub_;
+  ros::Publisher msg_pub_;
+
+  boost::shared_ptr<image_transport::ImageTransport> it_;
+
+  typedef line_segment_detector::LineSegmentDetectorConfig Config;
+  typedef dynamic_reconfigure::Server<Config> ReconfigureServer;
+  Config config_;
+  boost::shared_ptr<ReconfigureServer> reconfigure_server_;
+
+  bool debug_view_;
+  ros::Time prev_stamp_;
+
+  int lsd_refine_;
+  double lsd_scale_;
+  double lsd_sigma_scale_;
+  double lsd_quant_;
+  double lsd_angle_threshold_;
+  double lsd_log_eps_;
+  double lsd_density_threshold_;
+  int lsd_n_bins_;
+  double lsd_line_length_threshold_;
+
+  std::string window_name_;
+  static bool need_config_update_;
+
+  void reconfigureCallback(Config &new_config, uint32_t level)
+  {
+    config_ = new_config;
+
+    lsd_refine_ = config_.lsd_refine_type;
+    lsd_scale_ = config_.lsd_scale;
+    lsd_sigma_scale_ = config_.lsd_sigma_scale;
+    lsd_angle_threshold_ = config_.lsd_angle_threshold;
+    lsd_log_eps_ = config_.lsd_log_eps;
+    lsd_density_threshold_ = config_.lsd_density_threshold;
+    lsd_n_bins_ = config_.lsd_n_bins;
+    lsd_line_length_threshold_ = config_.lsd_line_length_threshold;
+  }
+
+  const std::string &frameWithDefault(const std::string &frame, const std::string &image_frame)
+  {
+    if (frame.empty())
+      return image_frame;
+    return frame;
+  }
+
+  void imageCallbackWithInfo(const sensor_msgs::ImageConstPtr& msg, const sensor_msgs::CameraInfoConstPtr& cam_info)
+  {
+    do_work(msg, cam_info->header.frame_id);
+  }
+
+  void imageCallback(const sensor_msgs::ImageConstPtr& msg)
+  {
+    do_work(msg, msg->header.frame_id);
+  }
+
+  static void trackbarCallback( int, void* )
+  {
+    need_config_update_ = true;
+  }
+
+  void do_work(const sensor_msgs::ImageConstPtr& msg, const std::string input_frame_from_msg)
+  {
+    // Work on the image.
+    try
+    {
+      // Convert the image into something opencv can handle.
+      cv::Mat frame = cv_bridge::toCvShare(msg, sensor_msgs::image_encodings::BGR8)->image;
+
+      // Do the work
+      cv::Mat src_gray;
+      /// Convert it to gray
+      if ( frame.channels() > 1 ) {
+        cv::cvtColor( frame, src_gray, cv::COLOR_RGB2GRAY );
+      } else {
+        src_gray = frame;
+      }
+
+      /// Create window
+      if( debug_view_) {
+        cv::namedWindow( window_name_, cv::WINDOW_AUTOSIZE );
+      }
+
+      std::string new_window_name;
+      cv::Mat grad;
+
+      cv::Ptr<cv::LineSegmentDetector> lsd = cv::createLineSegmentDetector(lsd_refine_, lsd_scale_,
```

Comment:
we don't have to create it every loop.
re-create instance when rosparam changed.
https://github.com/ros-perception/opencv_apps/pull/35/files#diff-51d9a14bcb3567cf07467833b23d4bebR91